File: python/03_geometry/main.py
"""03_geometry: 幾何変換 (resize / rotate / affine / perspective)
Usage (CLI): python main.py [input_image]
"""
import argparse
import sys
from pathlib import Path

import cv2
import numpy as np

sys.path.insert(0, str(Path(__file__).parent.parent))
from utils import DATA_DIR, load_image, show


def run(in_path: str | Path):
    src = load_image(in_path)
    h, w = src.shape[:2]

    # 1) リサイズ (半分 / 2倍)
    half = cv2.resize(src, None, fx=0.5, fy=0.5, interpolation=cv2.INTER_AREA)
    double = cv2.resize(src, None, fx=2.0, fy=2.0, interpolation=cv2.INTER_LINEAR)

    # 2) 回転 (中心 30°, スケール 1.0)
    center = (w / 2, h / 2)
    M_rot = cv2.getRotationMatrix2D(center, 30.0, 1.0)
    rotated = cv2.warpAffine(src, M_rot, (w, h))

    # 3) アフィン変換 (3点対応)
    src_pts = np.float32([[0, 0], [w - 1, 0], [0, h - 1]])
    dst_pts = np.float32([
        [w * 0.1, h * 0.2],
        [w * 0.9, h * 0.1],
        [w * 0.2, h * 0.9],
    ])
    M_aff = cv2.getAffineTransform(src_pts, dst_pts)
    affined = cv2.warpAffine(src, M_aff, (w, h))

    # 4) パースペクティブ変換 (せん断 + スケール)
    H = np.array([
        [1.0, 0.2, 0.0],
        [0.1, 1.0, 0.0],
        [0.0, 0.0, 1.0],
    ], dtype=np.float64)
    warped = cv2.warpPerspective(src, H, (w, h))

    # 5) 反転 (左右・上下・両方)
    flip_h = cv2.flip(src, 1)   # 左右
    flip_v = cv2.flip(src, 0)   # 上下

    print(f"元サイズ : {w} x {h}")
    print(f"半分     : {half.shape[1]} x {half.shape[0]}")
    print(f"ホモグラフィ H:\n{H}")

    pairs = [
        ("src", src),
        ("resize x0.5", half),
        ("rotate 30°", rotated),
        ("affine (3pt)", affined),
        ("perspective (shear)", warped),
        ("flip horizontal", flip_h),
        ("flip vertical", flip_v),
    ]
    return pairs


def main():
    parser = argparse.ArgumentParser(description="03_geometry: 幾何変換")
    parser.add_argument("input", nargs="?", default=str(DATA_DIR / "lena.png"))
    parser.add_argument("--no-show", action="store_true")
    args = parser.parse_args()

    pairs = run(args.input)
    if not args.no_show:
        show(pairs, cols=3)


if __name__ == "__main__":
    main()

File: cpp/03_geometry/main.cpp
// 03_geometry: 幾何変換 (resize / rotate / affine / perspective)
// Eigen で 3x3 のホモグラフィ行列を組み立て cv::Mat に流し込む例も含む。
// Usage: ./03_geometry <input_image>
#include <iostream>
#include <opencv2/opencv.hpp>
#include <Eigen/Dense>
#include "common/utils.hpp"

int main(int argc, char** argv) {
    const std::string in_path = tutorial::arg_or(argc, argv, 1, "../data/lena.png");
    cv::Mat src = tutorial::load_image(in_path);
    const cv::Size sz = src.size();

    // 1) リサイズ (半分)
    cv::Mat half;
    cv::resize(src, half, cv::Size(), 0.5, 0.5, cv::INTER_AREA);

    // 2) 回転 (中心 30deg) - getRotationMatrix2D は 2x3 アフィン行列
    cv::Point2f center(sz.width / 2.f, sz.height / 2.f);
    cv::Mat M_rot = cv::getRotationMatrix2D(center, 30.0, 1.0);
    cv::Mat rotated;
    cv::warpAffine(src, rotated, M_rot, sz);

    // 3) アフィン (3 点対応)
    cv::Point2f src_pts[3] = {{0, 0}, {sz.width - 1.f, 0}, {0, sz.height - 1.f}};
    cv::Point2f dst_pts[3] = {{sz.width * 0.1f, sz.height * 0.2f},
                              {sz.width * 0.9f, sz.height * 0.1f},
                              {sz.width * 0.2f, sz.height * 0.9f}};
    cv::Mat M_aff = cv::getAffineTransform(src_pts, dst_pts);
    cv::Mat affined;
    cv::warpAffine(src, affined, M_aff, sz);

    // 4) Eigen で 3x3 ホモグラフィ行列を組み、cv::warpPerspective に渡す
    //    例: 単純なせん断 + スケール
    Eigen::Matrix3d H_eig;
    H_eig << 1.0, 0.2, 0.0,
             0.1, 1.0, 0.0,
             0.0, 0.0, 1.0;
    cv::Mat H(3, 3, CV_64F);
    for (int r = 0; r < 3; ++r)
        for (int c = 0; c < 3; ++c) H.at<double>(r, c) = H_eig(r, c);

    cv::Mat warped;
    cv::warpPerspective(src, warped, H, sz);

    std::cout << "Eigen H =\n" << H_eig << std::endl;

    tutorial::show_and_wait({
        {"src", src},
        {"resize x0.5", half},
        {"rotate 30deg", rotated},
        {"affine 3pt", affined},
        {"perspective (Eigen H)", warped},
    });
    return 0;
}
